src/protocol/redpine_cc2500.c

```diff
@@ -0,0 +1,429 @@
+    s32 chan_val;
+
+    chan_val = Channels[chan];
+    chan_val = chan_val * STICK_SCALE / CHAN_MAX_VALUE + 1024;
+
+    if (chan_val > 2046)   chan_val = 2046;
+    else if (chan_val < 1) chan_val = 1;
+
+    return chan_val;
+}
+
+
+#define GET_FLAG(ch, mask) (Channels[ch] > 0 ? mask : 0)
+
+// For code readability
+enum {
+    CHANNEL1 = 0,
+    CHANNEL2,
+    CHANNEL3,
+    CHANNEL4,
+    CHANNEL5,
+    CHANNEL6,
+    CHANNEL7,
+    CHANNEL8,
+    CHANNEL9,
+    CHANNEL10,
+    CHANNEL11,
+    CHANNEL12,
+    CHANNEL13,
+    CHANNEL14,
+    CHANNEL15,    
+    CHANNEL16
+};
+
+
+static void redpine_data_frame() {
+    u16 chan[4];
+ 
+    packet_size = Model.proto_opts[PROTO_OPTS_PACKETSIZE];
+
+    memset(&packet[0], 0, packet_size);
+
+    packet[0] = packet_size-1;
+    packet[1] = fixed_id;
+    packet[2] = fixed_id >> 8;
+
+    chan[0] = scaleForPXX(0);
+    chan[1] = scaleForPXX(1);
+    chan[2] = scaleForPXX(2);
+    chan[3] = scaleForPXX(3);
+
+    packet[3] = chan[0];
+    packet[4] = (((chan[0] >> 8) & 0x07) | (chan[1] << 4)) | GET_FLAG(CHANNEL5, 0x08);
+    packet[5] = ((chan[1] >> 4) & 0x7F) | GET_FLAG(CHANNEL6, 0x80);
+    packet[6] = chan[2];
+    packet[7] = (((chan[2] >> 8) & 0x07) | (chan[3] << 4))  | GET_FLAG(CHANNEL7, 0x08); 
+    packet[8] = ((chan[3] >> 4) & 0x7F) | GET_FLAG(CHANNEL8, 0x80);
+    packet[9] = GET_FLAG(CHANNEL9, 0x01)
+            | GET_FLAG(CHANNEL10, 0x02)
+            | GET_FLAG(CHANNEL11, 0x04)
+            | GET_FLAG(CHANNEL12, 0x08)
+            | GET_FLAG(CHANNEL13, 0x10)
+            | GET_FLAG(CHANNEL14, 0x20)
+            | GET_FLAG(CHANNEL15, 0x40)
+            | GET_FLAG(CHANNEL16, 0x80);
+            
+    packet[10] = Model.proto_opts[PROTO_OPTS_LOOPTIME]; 
+    packet[11] = mixer_runtime/10;
+
+    u16 lcrc = frsky_crc(&packet[0], 12);
+    packet[12] = lcrc >> 8;
+    packet[13] = lcrc;
+
+}
+
+static u16 redpine_cb() {
+
+  switch(state) {
+    default:
+        set_start(47);
+        CC2500_SetPower(Model.tx_power);
+        CC2500_Strobe(CC2500_SFRX);
+        redpine_build_bind_packet();
+        CC2500_Strobe(CC2500_SIDLE);
+        CC2500_WriteData(packet, packet[0]+1);
+        state++;
+#ifndef EMULATOR
+        return 9000;
+#else
+        return 90;
+#endif
+    case REDPINE_BIND_DONE:
+        PROTOCOL_SetBindState(0);
+        initialize_data(0);
+        channr = 0;
+        state++;
+break;
+
+    case REDPINE_DATAM:
+#ifndef EMULATOR
+        CLOCK_RunMixer();    // clears mixer_sync, which is then set when mixer update complete
+        state = REDPINE_DATA1;
+        return mixer_runtime;
+#else
+        return 5;
+#endif
+          
+    case REDPINE_DATA1:
+        if (fine != (s8)Model.proto_opts[PROTO_OPTS_FREQFINE]) {
+            fine = (s8)Model.proto_opts[PROTO_OPTS_FREQFINE];
+            CC2500_WriteReg(CC2500_0C_FSCTRL0, fine);
+        }
+        CC2500_SetTxRxMode(TX_EN);
+        set_start(channr);
+        CC2500_SetPower(Model.tx_power);
```

Comment:
you don't want to set power if it is not changed.

Reply:
I missed this comment before - Most of the other protocols are doing similar.  Would it make sense to add a variable in CC2500_SetPower in cc2500.c to keep track of the last set power and only send it if it changed?

Reply:
I think that can be discussed. my option is that cc2500 is just an adopter to translate the command to wire protocol. it should not maintain any state. I would prefer you just copy how other protocol is doing for now.

The right solution should be creating a base protocol to share such logic instead of pushing down this to RF module driver. I had some ideas and if you like to we can discuss.